program_intent_engagement/apps/api/v1/views.py

```diff
@@ -1 +1,88 @@
-# Create your views here.
+"""
+V1 API Views
+"""
+import logging
+
+from edx_rest_framework_extensions.auth.jwt.authentication import JwtAuthentication
+from rest_framework import status
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from program_intent_engagement.apps.api.serializers import ProgramIntentSerializer
+from program_intent_engagement.apps.core.models import ProgramIntent, User
+
+log = logging.getLogger(__name__)
+
+
+class ProgramIntentsView(APIView):
+    # todo: add expected program intent data
+    """
+    View to insert program intents.
+
+    Given a program intent, this view will either insert the program intent. If the given program intent already
+    exists, no action is taken (no duplicate row).
+
+    Path: /api/[version]/program_intents
+
+    Accepts: [POST]
+
+    HTTP POST
+    Inserts a Program Intent.
+    Expected POST data:
+    POST data parameters:
+     * program_uuid: The uuid of the program intent.
+     * reason: The reason for how intent was measured.
+     * certainty: The certainty of the intent for the user.
+     * effective_timestamp: When the event occurred that created the intent.
+    Returns:
+     * 200: The program intent was inserted.
+     * 400: The program intent request was invalid.
+     * 409: Duplicate program intent was not inserted.
+    """
+
+    authentication_classes = (JwtAuthentication,)
+    # todo: do I need to add these? look into what these do
+    # serializer_class = ProgramIntentSerializer
+    # queryset = ProgramIntent.objects.all
+    # lookup_field = "user__lms_user_id"
+
+    @classmethod
+    def add_program_intent_user(cls, request):
+        # get internal user from the LMS user id in the request
+        user = User.objects.get(lms_user_id=request.user.lms_user_id)
+
+        # swap LMS user id for the internal user id before passing to serializer
+        request.data["user_id"] = user.id
+
+        return ProgramIntentSerializer(data=request.data)
+
+    def post(self, request):
+        """
+        Create and insert a program intent. Duplicates not inserted.
+        """
+        request_data = request.data
+
+        if request.user is None:
+            error_message = "Required field 'user' missing in request."
+            log.error(f"{error_message} Returning 400 BAD REQUEST")
+            return Response({"detail": error_message}, status=status.HTTP_400_BAD_REQUEST)
+
+        serializer = self.add_program_intent_user(request)
+
+        if serializer.is_valid():
+            # check if program intent exists, create if it does not exist
+            program_intent = ProgramIntent.objects.filter(**request_data)
+
+            if program_intent.exists():
+                response_status = status.HTTP_200_OK
+                data = {"detail": "This program intent already exists"}
+            # todo: figure out case if user changes their mind
```

Comment:
The [ticket](https://2u-internal.atlassian.net/browse/MST-1494) has this note: "What if there is a program intent that is the same except for effective date? I think we have to allow that to allow for users changing their mind."
I'm not sure I understand still what this means though? Would this mean that if there is a program intent with everything being the same (i.e. same program, same reason, same user) but the certainty being different? Or is this saying that if everything is the same except for the effective date, we would either override or keep both?

Reply:
That's a good question.

I am confused by the comment in the ticket. If you have a two program intents, and they differ by the `effective_timestamp`, then they are not duplicates, according to the set of fields we're using to check for uniqueness: `if there already exists a program intent matching user, program, reason, certainty and effective date DO NOT insert a duplicate row, we already know that.` They would both be inserted into the table anyway. In either case, `effective_timestamp` or `certainty`, they aren't duplicates, because they're different.

I know we had conversations about duplicates and clarifying this case, but I can't remember the outcome.

Reply:
when we talked about it we left it as insert one if the effective dates are different and maybe much later we could refuse duplicates whose effective dates are almost the same if that becomes a problem. So no more code is required here.

Reply:
Sweet, that makes sense. Thanks @MichaelRoytman and @ashultz0!